database-plugins/src/main/java/co/cask/hydrator/plugin/db/batch/source/DBSource.java

```diff
@@ -66,6 +85,135 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+      if (e instanceof SQLSyntaxErrorException) {
+        throw new BadRequestException(e.getMessage());
+      }
+      throw Throwables.propagate(e);
+    } finally {
+      DBUtils.deregisterShim(driverShim);
+    }
+  }
+
+  /**
+   * Endpoint method to get the output schema of a query.
+   * @param request {@link GetSchemaRequest} containing information required for connection and query to execute.
+   * @param pluginContext context to create plugins
+   * @return Map of field name to field type.
+   * @throws SQLException
+   * @throws InstantiationException
+   * @throws IllegalAccessException
+   * @throws BadRequestException If executing the query threw {@link SQLSyntaxErrorException}
+   * we get the message and throw it as BadRequestException
+   */
+  @Path("getSchema")
+  public Map<String, String> getSchema(GetSchemaRequest request,
+                                       EndpointPluginContext pluginContext)
+    throws SQLException, InstantiationException, IllegalAccessException, BadRequestException {
+
+    JDBCDriverShim driverShim = loadPluginClassAndGetDriver(request, pluginContext);
+    try {
+      Map<String, String> resultSchema = new LinkedHashMap<>();
+      try (Connection connection = getConnection(request.connectionString, request.user, request.password);) {
+        Statement statement = connection.createStatement();
+        String query  = request.importQuery;
+        statement.setMaxRows(1);
+        ResultSetMetaData resultSetMetaData = statement.executeQuery(query).getMetaData();
+
+        // column index starts from 1 in ResultSetMetaData
+        for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+          resultSchema.put(resultSetMetaData.getColumnName(i), resultSetMetaData.getColumnTypeName(i));
+        }
+      }
+      return resultSchema;
+    } catch (SQLException e) {
+      LOG.error("Exception while trying to get schema in a database for query {}.",
+                request.importQuery, e);
+
+      if (e instanceof SQLSyntaxErrorException) {
+        throw new BadRequestException(e.getMessage());
+      }
+      throw Throwables.propagate(e);
+    } finally {
+      DBUtils.deregisterShim(driverShim);
+    }
+  }
+
+  private JDBCDriverShim loadPluginClassAndGetDriver(ListTablesRequest request, EndpointPluginContext pluginContext)
+    throws IllegalAccessException, InstantiationException, SQLException {
+    Class<? extends Driver> driverClass =
+      pluginContext.loadPluginClass(request.jdbcPluginType, request.jdbcPluginName, PluginProperties.builder().build());
+
+    try {
+      return DBUtils.ensureJDBCDriverIsAvailable(driverClass, request.connectionString,
+                                                 request.jdbcPluginType, request.jdbcPluginName);
+    } catch (IllegalAccessException | InstantiationException | SQLException e) {
+      LOG.error("Unable to load or register JDBC driver {} while getting the tables in a database.",
```

Comment:
in general, log then propagate is a sign of error handling happening in the wrong place.  This method says it throws these exceptions already, we should not wrap them in a runtime exception.

This error message is also wrong, its not getting database tables.